lib/modem_attest_token/modem_attest_token.c

```diff
@@ -0,0 +1,494 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <string.h>
+#include <zephyr.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <modem/nrf_modem_lib.h>
+#include <modem/at_cmd.h>
+#include <sys/base64.h>
+#include <tinycbor/cbor.h>
+#include <tinycbor/cbor_buf_reader.h>
+#include <modem/modem_attest_token.h>
+
+#define GET_BASE64_LEN(n) (((4 * n / 3) + 3) & ~3)
+#define BASE64_PAD_CHAR '='
+#define AT_ATTEST_CMD "AT%ATTESTTOKEN"
+
+/* Attestation message format:
+ * Tag (55799)
+ * Array (len 5)
+ * - INT: msg type (NRF_ID_SRVC_MSG_TYPE_ID_V1)
+ * - BSTR: device UUID
+ * - INT: device type
+ * - BSTR: FW UUID
+ * - BSTR: nonce
+ */
+#define ATTEST_MAP_LEN 5
+
+#define ATTEST_TOKEN_RESP_LEN 200
+BUILD_ASSERT(
+	ATTEST_TOKEN_RESP_LEN <= CONFIG_AT_CMD_RESPONSE_MAX_LEN,
+	"Max AT cmd response length is too small for attestation token response");
+
+static int parse_resp(char *const resp, char **attest, char **cose)
+{
+	if (!resp || !attest || !cose) {
+		return -EINVAL;
+	}
+
+	char *cose_end = NULL;
+	char *attest_end = NULL;
+
+	*cose = NULL;
+
+	*attest = strchr(resp, '"');
+	if (!(*attest)) {
+		return -EBADMSG;
+	}
+	/* Move beyond first quotation mark */
+	*attest = *attest + 1;
+
+	attest_end = strchr(*attest, '.');
+	if (!(*attest_end)) {
+		return -EBADMSG;
+	}
+
+	/* Replace . with a NULL */
+	*attest_end = '\0';
+
+	*cose = attest_end + 1;
+	cose_end = strchr(*cose, '"');
+	if (!cose_end) {
+		return -EBADMSG;
+	}
+
+	/* Replace " with a NULL */
+	*cose_end = '\0';
+
+	return 0;
+}
+
+int modem_attest_token_get(struct nrf_attestation_token *const token)
+{
+	if (!token) {
+		return -EINVAL;
+	}
+
+	int ret = 0;
+	enum at_cmd_state state;
+	char *cmd_resp = NULL;
+	char *attest = NULL;
+	char *cose = NULL;
+	size_t sz;
+
+	token->attest = NULL;
+	token->cose = NULL;
+
+	/* Init required modules */
+	ret = nrf_modem_lib_init(NORMAL_MODE);
```

Comment:
The carrier lib docs state that the application should not call this function when used together with the library. 
Perhaps one could try `at_cmd_init()` first and fall back to try to initialize both libraries on error?

Reply:
> `at_cmd_init` just returns zero if it is already initialized.

`at_cmd_init` attempts to open a socket, so if that returns zero, it should indicate that modem lib is already initialized successfully.
I looked at `nrf_modem_lib_init` and couldn't see protection against repeated initializations, but maybe that's fine. @lemrey do you have any best practices wrt this?

Reply:
`at_cmd_init` returns 0 on success, which includes subsequent calls... maybe i'm misunderstanding?

Reply:
So here I would suggest to not initialize other libraries at all, instead document the dependency and let the user initialize them when appropriate. In general, I never do this because it introduces an alternative pattern for initializing things, plus it is "hidden" or implicit.